File: algorithms/python/productExceptSelf/productExceptSelf.py
"""
Source	: https://leetcode.com/problems/product-of-array-except-self/
Name	: Matthew Hernandez
Date	: 2025-06-04

Prefix & Suffix solution with space complexity O(n)
– Not the most optimal solution but intuitive (besides brute-force)
"""

class Solution(object):
	def productExceptSelf(self, nums):
		"""
		:type nums: List[int]
		:rtype: List[int]
		"""
		n = len(nums)
		prefix = [1] * n
		suffix = [1] * n
		result = [1] * n

		# Fill prefix array
		for i in range(1, n):
			prefix[i] = prefix[i - 1] * nums[i - 1]

		# Fill suffix array
		for i in range(n - 2, -1, -1):
			suffix[i] = suffix[i + 1] * nums[i + 1]

		# Fill product array
		for i in range(0, n):
			result[i] = prefix[i] * suffix[i]

		return result

File: leetcode/java/productExceptSelf/productExceptSelf.java
// Source      : https://leetcode.com/problems/product-of-array-except-self/
// Author      : Matthew Hernandez
// Date        : 2025-06-6
// Algorithm   : Prefix, Postfix arrays

import java.util.Arrays;


class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        Arrays.fill(prefix, 1);
        int[] postfix = prefix.clone();
        int[] res = new int[n];

        // Fill prefix array
        for (int i=1; i < nums.length;i++) {
            prefix[i] = prefix[i-1] * nums[i-1];
        }
        // Fill postfix array
        for (int i=nums.length-2; i > -1;i--){
            postfix[i] = postfix[i+1] * nums[i+1];
        }
        // Fill product array
        for (int i=0; i < nums.length;i++) {
            res[i] = prefix[i] * postfix[i];
        }
        return res;
        // return new int[]{};
        
    }
}